Lib/imaplib.py

```diff
@@ -1049,7 +1048,17 @@ def _append_untagged(self, typ, dat):
 
         # During idle, queue untagged responses for delivery via iteration
         if self._idle_capture:
-            self._idle_responses.append((typ, dat))
+            # Responses containing literal strings are passed to us one data
+            # fragment at a time, while others arrive in a single call.
+            if (not self._idle_responses or
+                isinstance(self._idle_responses[-1][1][-1], bytes)):
```

Comment:
This logic looks suspect. I think the isinstance() checks whether the last data fragment of the last received reponse is bytes, and if so, we *are* continuing a fragmented response. Maybe you meant `not isinstance(...)`?

Rembember that `not` binds tighter than `or`.

I would add tests for all paths through this logic (unless I've got it all wrong).

Reply:
Oh, I see. Please ignore me -- I don't know the first thing about IMAP so I was just guessing. If a bytes string indicates the last part of a response, then the logic looks correct. Sorry for the false alarm!

I have pinged Serhiy off-line, if I don't hear from him soon I can merge, assuming you and your collaborators are happy with the current state.

Reply:
> Shall I ping you when I'm comfortable merging?

Sure.

On declaring the interface as provisional, that's usually only done if there are issues in the *interface* that we're not sure about. (And IIRC the SC has decided that at least for PEPs, "provisional" should not be used any more.)

Until beta 1 we can easily revert these changes, though I bet if you wait half a year it's going to be tricky due to other minor changes happening in the meantime. Even during the beta phase, if we find a big enough problem with either the interface or the implementation, we might be able to roll it back (we had a major change to the GC implementation rolled back during the rc phase for 3.13). But it's a lot more acceptable to *fix* issues brought up during the beta release, since that is the general intent of the beta phase -- shake out the bugs in new implementations and interfaces.

Are you worried that implementation changes made to support IDLE could break existing uses of imaplib (without using IDLE)? That's something to be careful about, and it might show up during the beta cycle -- at that point you have time to fix it, even if the fix is a bit baroque (that's routine for the stdlib). You might try to find existing users and see if they're willing to test with an early beta.

Or are you worried that the implementation of IDLE might be broken? Again, if it's broken it can be fixed, and it's not the end of the world if a fix must be applied after the 3.14.0 release (but not an excuse for being sloppy now, of course). Traditionally we're a bit more lax for brand new features.

OTOH, if you have your doubts about the new *interface*, you may want to sit back and think everything through. The only area where I have some doubts myself is the warnings about IMAP4_stream on Windows. Would it be better to just raise an exception if that combination is detected and idle() is called? Relying on observed behavior of one particular IMAP server seems a little iffy. We have plenty of features (or feature combinations) in the stdlib that are restricted to certain platforms. We could later lift that restriction by doing some Windows-specific magic (I'm sure that there's a way to use a native Windows select-like API on pipes, which is what IMAP_stream seems to be using.) I recommend not attempting that Windows hackery in this PR.

Reply:
> I wonder how hard it will be to find existing imaplib users to test the beta. A quick search in the Discourse forum doesn't turn up many hits. Is the email list gone? Is there another channel that reaches a lot of folks beyond the contributor community?

You could post a call to python-announce@python.org, which is still a mailing list AFAIK (but not as heavily used as in the past). You could search GitHub for projects that import imaplib and contact them through their tracker or listed contact (if any).

I think discuss.python.org has some widely-read lists where you could ask who uses imaplib.